File: sigzenjira/sigzenjira/page/project_billing/project_billing.py
import json

import frappe
from frappe import _
from frappe.utils import flt

from sigzenjira.events.task import WORK_ITEM_TYPE_PRIVILEGED_ROLES
from sigzenjira.events.timesheet import recompute_actual_time, recompute_project_billable_hours
from sigzenjira.permission.project import user_is_project_member

# Director, Product Owner, Projects Manager - and System Manager, which is how
# this app spells "administrator" everywhere else. Reading and adjusting are one
# gate, not two: the page exists to review what gets invoiced, and a project's
# billable figures are not something a Projects User has any business browsing.
#
# Reusing the events/billable.py set is the point. Deciding how many of an
# hour's minutes reach the invoice is the same money decision as deciding
# whether the work is billable at all, so the two must never drift apart.
#
# Must stay in step with the `roles` list in project_billing.json: frappe's
# Page.is_permitted is a plain role intersection with no System Manager bypass,
# so that file is a separate gate and neither one implies the other.
BILLING_ROLES = WORK_ITEM_TYPE_PRIVILEGED_ROLES

# Emission order for siblings. Follows the hierarchy's own shape; anything
# unclassified sorts last so a legacy Task with no work item type never appears
# above the structured work.
WORK_ITEM_ORDER = {"Epic": 0, "Story": 1, "Task": 2, "Sub-task": 3}

TASK_FIELDS = [
	"name",
	"subject",
	"parent_task",
	"custom_task_work_item_type",
	"custom_task_is_billable",
	"custom_task_billable_hours",
	"custom_task_non_billable_hours",
	"actual_time",
]

DETAIL_FIELDS = [
	"name",
	"parent",
	"task",
	"project",
	"activity_type",
	"from_time",
	"hours",
	"billing_hours",
	"custom_timesheet_detail_non_billable_hours",
	"custom_timesheet_detail_mark_as_read",
	"is_billable",
	"sales_invoice",
]


def _guard_billing_access():
	if not BILLING_ROLES & set(frappe.get_roles(frappe.session.user)):
		frappe.throw(
			_("Only a Director, Product Owner, or Projects Manager can review billable hours."),
			frappe.PermissionError,
		)


def _guard_project_access(project):
	# Every role in BILLING_ROLES is also in PROJECT_SCOPE_BYPASS_ROLES, so today
	# this never refuses anyone who got past _guard_billing_access. Kept anyway:
	# it is the same record-level rule the Task and Project list views run on
	# (permission/project.py), so if BILLING_ROLES is ever widened to a
	# project-scoped role, these whitelisted methods stay scoped with it instead
	# of quietly becoming the one way to read any project in the company.
	if not user_is_project_member(project):
		frappe.throw(_("Not permitted"), frappe.PermissionError)


@frappe.whitelist()
def get_bootstrap():
	_guard_billing_access()

	# Billable projects only: on a non-billable Project every row is unbilled by
	# construction (the clamp in events/billable.py forbids anything else), so
	# there is nothing on one to review or adjust.
	#
	# get_list, NOT get_all: frappe.get_all forces ignore_permissions=True, which
	# would hand a project-scoped user the whole company's project list. get_list
	# runs project_query_conditions (permission/project.py), so this is already
	# scoped to the Projects the user is a Project User on.
	projects = frappe.get_list(
		"Project",
		filters={"custom_project_is_billable": 1},
		fields=["name", "project_name", "company"],
		order_by="project_name asc",
		limit_page_length=0,
	)

	# Served rather than hardcoded in the JS so the filter's options and the
	# tree's sort order can never drift apart.
	return {"projects": projects, "work_item_types": list(WORK_ITEM_ORDER)}


def _timesheet_rows(project, from_date=None, to_date=None, employee=None):
	filters = {"project": project, "docstatus": 1}
	if from_date:
		filters["from_time"] = [">=", from_date]
	if to_date:
		# to_time rather than from_time so a row that starts on the last day of
		# the range is not dropped for finishing after midnight.
		filters["to_time"] = ["<=", f"{to_date} 23:59:59"]

	rows = frappe.get_all(
		"Timesheet Detail",
		filters=filters,
		fields=DETAIL_FIELDS,
		order_by="from_time asc",
		limit_page_length=0,
	)

	# Who logged the time lives on the parent Timesheet, not the row, so it takes
	# one lookup keyed on the parents actually in view.
	parents = {row.parent for row in rows if row.parent}
	sheets = (
		frappe.get_all(
			"Timesheet",
			filters={"name": ["in", list(parents)]},
			fields=["name", "employee", "employee_name"],
			limit_page_length=0,
		)
		if parents
		else []
	)
	by_sheet = {sheet.name: sheet for sheet in sheets}

	for row in rows:
		sheet = by_sheet.get(row.parent)
		row.employee = sheet.employee if sheet else None
		row.employee_name = (sheet.employee_name or sheet.employee) if sheet else None

	# The dropdown is built from who actually billed hours here, NOT from
	# Employee.company: a person employed by one company can log time on another
	# company's project, and scoping this by company would hide their hours from
	# the invoice review while still counting them in the project total.
	#
	# Built before the employee filter is applied, so selecting someone does not
	# collapse the list to just them.
	employees = sorted(
		{(row.employee, row.employee_name) for row in rows if row.employee},
		key=lambda pair: pair[1] or pair[0],
	)

	if employee:
		rows = [row for row in rows if row.employee == employee]

	return rows, [{"name": name, "employee_name": label} for name, label in employees]


def _detail_row(row, depth):
	return {
		"row_type": "timesheet",
		"indent": depth,
		"name": row.name,
		"timesheet": row.parent,
		"label": row.activity_type or _("Time log"),
		"from_time": row.from_time,
		"hours": flt(row.hours),
		"billing_hours": flt(row.billing_hours),
		"non_billable_hours": flt(row.custom_timesheet_detail_non_billable_hours),
		"is_billable": row.is_billable,
		"employee": row.employee,
		"employee_name": row.employee_name,
		"mark_as_read": row.custom_timesheet_detail_mark_as_read,
		# An invoiced row is frozen: changing what was already billed to a
		# customer is an invoice correction, not a timesheet edit.
		"locked": bool(row.sales_invoice),
		"sales_invoice": row.sales_invoice,
	}


def _task_row(task, depth):
	return {
		"row_type": "task",
		"indent": depth,
		"name": task.name,
		"label": task.subject,
		"work_item_type": task.custom_task_work_item_type,
		"is_billable": task.custom_task_is_billable,
		# The stored rollup, not a sum of the rows below: on an Epic or Story
		# these are the whole subtree's hours, which the rows under that node
		# alone would understate.
		#
		# ponytail: a work item's figures are all-time and ignore the
		# date/employee filters. Only the timesheet rows and the "in view" totals
		# narrow. Recompute these from the filtered rows if the mismatch
		# confuses anyone.
		"hours": flt(task.actual_time),
		"billing_hours": flt(task.custom_task_billable_hours),
		"non_billable_hours": flt(task.custom_task_non_billable_hours),
	}


def _build_tree(tasks, rows_by_task, work_item_type=None):
	names = {task.name for task in tasks}
	children = {}
	for task in tasks:
		# A parent outside this project (or deleted) would make its whole subtree
		# unreachable from the root walk, so those tasks are promoted to roots
		# rather than silently dropped.
		parent = task.parent_task if task.parent_task in names else None
		children.setdefault(parent, []).append(task)

	for bucket in children.values():
		bucket.sort(key=lambda t: (WORK_ITEM_ORDER.get(t.custom_task_work_item_type, 99), t.name))

	if work_item_type:
		# Re-root rather than filter: every work item of the chosen type becomes a
		# top-level row and keeps its whole subtree, so picking Story gives one row
		# per Story with its Tasks, Sub-tasks and time logs still nested under it.
		# A Story whose Epic is now out of view is promoted, not dropped with it.
		roots = sorted((t for t in tasks if t.custom_task_work_item_type == work_item_type), key=lambda t: t.name)
	else:
		roots = children.get(None, [])

	# validate_hierarchy already forbids a parent_task cycle, but this walk would
	# recurse until the stack blew if one ever reached the table by another route
	# (a raw db.set_value, an import). One set is cheaper than that outage.
	seen = set()

	def build(task, depth):
		# The subtree's rows, or nothing at all when no time was logged anywhere
		# under it. An empty branch is noise on a page about hours - and with a
		# date or employee filter on, it is most of the tree.
		if task.name in seen:
			return []
		seen.add(task.name)

		details = [_detail_row(row, depth + 1) for row in rows_by_task.get(task.name, [])]

		below = []
		for child in children.get(task.name, []):
			below += build(child, depth + 1)

		if not details and not below:
			return []

		return [_task_row(task, depth), *details, *below]

	out = []
	for root in roots:
		out += build(root, 0)
	return out


@frappe.whitelist()
def get_project_billing(project, from_date=None, to_date=None, employee=None, work_item_type=None):
	_guard_billing_access()
	_guard_project_access(project)

	tasks = frappe.get_all(
		"Task", filters={"project": project}, fields=TASK_FIELDS, limit_page_length=0
	)
	details, employees = _timesheet_rows(project, from_date, to_date, employee)

	rows_by_task = {}
	unassigned = []
	for row in details:
		if row.task:
			rows_by_task.setdefault(row.task, []).append(row)
		else:
			unassigned.append(row)

	# An unrecognised type would silently empty the page, so it is dropped rather
	# than passed through to a walk that can only match nothing.
	rows = _build_tree(tasks, rows_by_task, work_item_type if work_item_type in WORK_ITEM_ORDER else None)

	# Time logged against the Project with no Task at all. No Task rollup can see
	# it, but recompute_project_billable_hours counts it, so leaving it off the
	# page would make the project total look wrong. It belongs to no work item, so
	# a work item type filter is exactly when it should not appear.
	if unassigned and not work_item_type:
		rows.append(
			{
				"row_type": "task",
				"indent": 0,
				"name": "",
				"label": _("Time logged without a Task"),
				"work_item_type": "",
				"is_billable": 0,
				"hours": sum(flt(row.hours) for row in unassigned),
				"billing_hours": sum(flt(row.billing_hours) for row in unassigned),
				"non_billable_hours": sum(
					flt(row.custom_timesheet_detail_non_billable_hours) for row in unassigned
				),
			}
		)
		rows += [_detail_row(row, 1) for row in unassigned]

	# Totalled off the timesheet rows actually emitted, not off the Project's
	# stored fields: those always cover the project's whole life, and would
	# contradict the rows on screen the moment a filter is applied. Summing the
	# emitted rows rather than the query result keeps "in view" literal - a work
	# item type filter drops the no-task bucket, and the total drops with it.
	in_view = [row for row in rows if row["row_type"] == "timesheet"]

	return {
		"rows": rows,
		"employees": employees,
		"totals": {
			"hours": sum(flt(row["hours"]) for row in in_view),
			"billing_hours": sum(flt(row["billing_hours"]) for row in in_view),
			"non_billable_hours": sum(flt(row["non_billable_hours"]) for row in in_view),
		},
		"project_totals": frappe.db.get_value(
			"Project",
			project,
			["custom_project_billable_hours", "custom_project_non_billable_hours"],
			as_dict=True,
		),
	}


def _validated_updates(updates):
	by_name = {row["name"]: flt(row["billing_hours"]) for row in updates}
	rows = frappe.get_all(
		"Timesheet Detail",
		filters={"name": ["in", list(by_name)], "docstatus": 1},
		fields=["name", "parent", "task", "project", "hours", "is_billable", "sales_invoice"],
		limit_page_length=0,
	)

	if len(rows) != len(by_name):
		frappe.throw(_("Some of those timesheet rows no longer exist. Refresh the page and try again."))

	for row in rows:
		_guard_project_access(row.project)
		new_hours = by_name[row.name]

		if row.sales_invoice:
			frappe.throw(
				_("Row on {0} is already invoiced ({1}) and cannot be changed here.").format(
					row.parent, row.sales_invoice
				)
			)
		if new_hours < 0:
			frappe.throw(_("Billable hours cannot be negative."))
		if new_hours > flt(row.hours):
			frappe.throw(
				_("Billable hours cannot exceed the {0} hours actually worked on this row.").format(
					flt(row.hours)
				)
			)
		if not row.is_billable and new_hours:
			frappe.throw(
				_("This row is not billable, so its billable hours must stay 0. Change the Task instead.")
			)

	return by_name, rows


def _apply_to_timesheet(timesheet_name, by_name):
	doc = frappe.get_doc("Timesheet", timesheet_name)
	changed = []

	for row in doc.time_logs:
		if row.name not in by_name:
			continue

		previous = flt(row.billing_hours)
		row.billing_hours = by_name[row.name]
		row.custom_timesheet_detail_non_billable_hours = flt(row.hours) - flt(row.billing_hours)
		# Core's own rate maths rather than a copy of it. update_cost only fills a
		# rate that is still 0, so a billing_rate someone set by hand survives -
		# it just recomputes the amounts off the new billing_hours.
		row.update_cost(doc.employee)

		frappe.db.set_value(
			"Timesheet Detail",
			row.name,
			{
				"billing_hours": row.billing_hours,
				"custom_timesheet_detail_non_billable_hours": row.custom_timesheet_detail_non_billable_hours,
				"billing_amount": row.billing_amount,
				"base_billing_amount": row.base_billing_amount,
			},
			update_modified=False,
		)
		changed.append(f"{row.idx}: {previous} -> {flt(row.billing_hours)}")

	# Core's Timesheet.on_update_after_submit does not call
	# calculate_total_amounts, so a billing_hours change made after submit leaves
	# the parent's own totals stale. Recomputed here off the rows just written.
	doc.calculate_total_amounts()
	doc.calculate_percentage_billed()
	frappe.db.set_value(
		"Timesheet",
		timesheet_name,
		{
			"total_billable_hours": doc.total_billable_hours,
			"total_billable_amount": doc.total_billable_amount,
			"base_total_billable_amount": doc.base_total_billable_amount,
			"total_billed_amount": doc.total_billed_amount,
			"base_total_billed_amount": doc.base_total_billed_amount,
			"total_billed_hours": doc.total_billed_hours,
			"per_billed": doc.per_billed,
		},
		update_modified=False,
	)

	# The trail. These edits land on a submitted document through db.set_value,
	# which writes no version history of its own, so without this a billed figure
	# could change with nothing anywhere recording who moved it or from what.
	if changed:
		doc.add_comment(
			"Info",
			_("Billable hours adjusted from Project Billing - row {0}").format(", ".join(changed)),
		)


@frappe.whitelist()
def update_billing_hours(updates):
	if isinstance(updates, str):
		updates = json.loads(updates)

	_guard_billing_access()

	if not updates:
		return {"updated": 0}

	by_name, rows = _validated_updates(updates)

	for timesheet_name in {row.parent for row in rows}:
		_apply_to_timesheet(timesheet_name, by_name)

	# Task rollups first: recompute_project_billable_hours reads the rows, not the
	# Tasks, so the order does not matter for correctness - but a Task left stale
	# would show a figure contradicting the Project total on the same screen.
	for task in {row.task for row in rows if row.task}:
		recompute_actual_time(task)
	for project in {row.project for row in rows if row.project}:
		recompute_project_billable_hours(project)

	return {"updated": len(by_name)}


@frappe.whitelist()
def set_mark_as_read(names, value):
	if isinstance(names, str):
		names = json.loads(names)

	_guard_billing_access()

	if not names:
		return {"updated": 0}

	rows = frappe.get_all(
		"Timesheet Detail",
		filters={"name": ["in", names]},
		fields=["name", "project"],
		limit_page_length=0,
	)
	for row in rows:
		_guard_project_access(row.project)

	for row in rows:
		frappe.db.set_value(
			"Timesheet Detail",
			row.name,
			"custom_timesheet_detail_mark_as_read",
			1 if int(value) else 0,
			update_modified=False,
		)

	return {"updated": len(rows)}

File: sigzenjira/sigzenjira/page/project_billing/project_billing.js
// Project Billing - review and adjust billable hours per project.
//
// The tree is Project > Epic > Story > Task > Sub-task > timesheet rows. Only
// the timesheet rows are editable: every level above them shows the stored
// rollup that recompute_actual_time maintains, so typing into one would be
// editing a total rather than a fact.

frappe.pages["project-billing"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Project Billing"),
		single_column: true,
	});

	new ProjectBilling(page);
};

class ProjectBilling {
	constructor(page) {
		this.page = page;
		this.rows = [];
		// fieldname -> pending billable-hours value. Nothing reaches the server
		// until Save, so a mistyped figure costs nothing until it is committed.
		this.pending = {};
		this.collapsed = new Set();
		// Every billable project the user may see, kept so the Company filter can
		// narrow the Project list without a second server round trip.
		this.all_projects = [];
		// Employees with submitted hours on the project currently in view. Filled
		// by each load, so the dropdown can only ever offer people who billed here.
		this.employees = [];

		this.make_filters();
		this.make_body();
		this.bootstrap();
	}

	make_filters() {
		// Company narrows the Project list and nothing else. It deliberately does
		// not reach the Employee filter: see set_employee_options.
		this.company_field = this.page.add_field({
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			change: () => this.set_project_options(),
		});

		this.project_field = this.page.add_field({
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Select",
			change: () => {
				// A person who billed the old project is unlikely to have billed
				// this one, and a stale value would silently filter every row away.
				this.employee_field.set_value("");
				this.refresh();
			},
		});

		// Re-roots the tree at the chosen level: pick Story and every Story is a
		// top-level row with its own Tasks, Sub-tasks and time logs still nested
		// and collapsible underneath it.
		this.work_item_type_field = this.page.add_field({
			fieldname: "work_item_type",
			label: __("Work Item Type"),
			fieldtype: "Select",
			change: () => this.refresh(),
		});

		this.employee_field = this.page.add_field({
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",
			get_query: () => {
				const names = this.employees.map((e) => e.name);
				// No project loaded yet, or nobody has billed it: offer nothing
				// rather than the whole Employee table.
				return { filters: { name: ["in", names.length ? names : [""]] } };
			},
			change: () => this.refresh(),
		});

		this.from_date_field = this.page.add_field({
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			change: () => this.refresh(),
		});

		this.to_date_field = this.page.add_field({
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			change: () => this.refresh(),
		});

		this.save_button = this.page.set_primary_action(__("Save Changes"), () => this.save());
		this.set_dirty(false);
	}

	make_body() {
		this.body = $('<div class="project-billing"></div>').appendTo(this.page.main);
	}

	bootstrap() {
		frappe.call({ method: this.method("get_bootstrap") }).then((r) => {
			const data = r.message || {};
			this.all_projects = data.projects || [];

			this.work_item_type_field.df.options = [{ value: "", label: __("Whole hierarchy") }].concat(
				(data.work_item_types || []).map((t) => ({ value: t, label: __(t) }))
			);
			this.work_item_type_field.refresh();

			if (!this.all_projects.length) {
				this.render_message(
					__("No billable projects available to you. Tick Billable on a Project first.")
				);
				return;
			}

			this.set_project_options();
		});
	}

	set_project_options() {
		const company = this.company_field.get_value();
		const matching = company
			? this.all_projects.filter((p) => p.company === company)
			: this.all_projects;

		const options = matching.map((p) => ({ value: p.name, label: p.project_name || p.name }));

		this.project_field.df.options = [{ value: "", label: __("Select a project") }].concat(options);
		this.project_field.refresh();

		// The project in view may not belong to the newly picked company. Clearing
		// it drives the project change handler, which reloads and drops the
		// employee too, so no separate refresh is needed here.
		const selected = this.project_field.get_value();
		if (selected && !matching.some((p) => p.name === selected)) {
			this.project_field.set_value("");
		}
	}

	set_employee_options(employees) {
		this.employees = employees || [];

		// Anyone who billed hours here stays selectable regardless of which company
		// employs them - a cross-company timesheet is still money on this invoice.
		const selected = this.employee_field.get_value();
		if (selected && !this.employees.some((e) => e.name === selected)) {
			this.employee_field.set_value("");
		}
	}

	method(name) {
		return `sigzenjira.sigzenjira.page.project_billing.project_billing.${name}`;
	}

	filters() {
		return {
			project: this.project_field.get_value(),
			work_item_type: this.work_item_type_field.get_value(),
			employee: this.employee_field.get_value(),
			from_date: this.from_date_field.get_value(),
			to_date: this.to_date_field.get_value(),
		};
	}

	refresh() {
		const filters = this.filters();
		if (!filters.project) {
			this.employees = [];
			this.render_message(__("Select a project to review its billable hours."));
			return;
		}

		// Discarded deliberately: the figures are about to be replaced, so keeping
		// edits keyed to rows that may not survive the reload would silently apply
		// them to a different set than the one they were typed against.
		this.pending = {};
		this.set_dirty(false);

		frappe.call({ method: this.method("get_project_billing"), args: filters }).then((r) => {
			const data = r.message || {};
			this.rows = data.rows || [];
			this.set_employee_options(data.employees);
			this.render(data);
		});
	}

	render_message(message) {
		this.body.html(`<div class="project-billing-empty">${frappe.utils.escape_html(message)}</div>`);
	}

	render(data) {
		if (!this.rows.length) {
			this.render_message(__("No submitted timesheet hours on this project for that period."));
			return;
		}

		const totals = data.totals || {};
		const project_totals = data.project_totals || {};

		this.body.html(`
			<div class="project-billing-summary">
				${this.summary_tile(__("Hours in view"), totals.hours)}
				${this.summary_tile(__("Billable in view"), totals.billing_hours)}
				${this.summary_tile(__("Non-billable in view"), totals.non_billable_hours)}
				${this.summary_tile(
					__("Project billable (all time)"),
					project_totals.custom_project_billable_hours
				)}
				${this.summary_tile(
					__("Project non-billable (all time)"),
					project_totals.custom_project_non_billable_hours
				)}
			</div>
			<div class="project-billing-table-wrapper">
				<table class="project-billing-table">
					<thead>
						<tr>
							<th class="pb-work">${__("Work Item / Time Log")}</th>
							<th class="pb-num">${__("Hours")}</th>
							<th class="pb-num">${__("Billable")}</th>
							<th class="pb-num">${__("Non-Billable")}</th>
							<th class="pb-read">${__("Read")}</th>
						</tr>
					</thead>
					<tbody></tbody>
				</table>
			</div>
		`);

		this.render_rows();
		this.bind_row_events();
	}

	summary_tile(label, value) {
		return `
			<div class="project-billing-tile">
				<div class="project-billing-tile-label">${frappe.utils.escape_html(label)}</div>
				<div class="project-billing-tile-value">${format_number(value || 0, null, 2)}</div>
			</div>
		`;
	}

	// A row is hidden when any task above it is collapsed. Tracked by walking the
	// flat list and remembering the shallowest collapsed depth still in force,
	// which is cheaper than rebuilding the tree client-side just to hide things.
	visible_rows() {
		let hidden_from = null;
		return this.rows.filter((row) => {
			if (hidden_from !== null && row.indent > hidden_from) return false;

			hidden_from = null;
			if (row.row_type === "task" && this.collapsed.has(row.name)) {
				hidden_from = row.indent;
			}
			return true;
		});
	}

	render_rows() {
		const body = this.body.find("tbody");
		body.empty();

		this.visible_rows().forEach((row) => {
			body.append(row.row_type === "task" ? this.task_row_html(row) : this.detail_row_html(row));
		});
	}

	task_row_html(row) {
		const has_children = this.has_children(row);
		const caret = has_children
			? `<span class="pb-caret">${this.collapsed.has(row.name) ? "&#9656;" : "&#9662;"}</span>`
			: '<span class="pb-caret pb-caret-empty"></span>';

		const badge = row.work_item_type
			? `<span class="pb-badge pb-badge-${row.work_item_type.toLowerCase().replace("-", "")}">${
					row.work_item_type
			  }</span>`
			: "";

		const link = row.name
			? `<a href="/app/task/${encodeURIComponent(row.name)}">${frappe.utils.escape_html(row.label)}</a>`
			: frappe.utils.escape_html(row.label);

		return `
			<tr class="pb-task-row" data-name="${frappe.utils.escape_html(row.name)}">
				<td class="pb-work" style="padding-left: ${12 + row.indent * 20}px">
					${caret}${badge}${link}
					${row.is_billable ? '<span class="pb-billable-dot" title="Billable"></span>' : ""}
				</td>
				<td class="pb-num">${format_number(row.hours, null, 2)}</td>
				<td class="pb-num">${format_number(row.billing_hours, null, 2)}</td>
				<td class="pb-num">${format_number(row.non_billable_hours, null, 2)}</td>
				<td class="pb-read"></td>
			</tr>
		`;
	}

	detail_row_html(row) {
		const pending = this.pending[row.name];
		const billing_hours = pending === undefined ? row.billing_hours : pending;
		const non_billable = flt(row.hours) - flt(billing_hours);
		// The page's only gate is BILLING_ROLES, checked server-side on open, so
		// anyone looking at this can edit. Only the row itself can freeze a cell.
		const editable = !row.locked && row.is_billable;

		const billing_cell = editable
			? `<input type="number" class="pb-input" step="0.01" min="0" max="${row.hours}"
					value="${format_number(billing_hours, null, 2)}" data-name="${row.name}">`
			: format_number(billing_hours, null, 2);

		const who = row.employee_name ? ` &middot; ${frappe.utils.escape_html(row.employee_name)}` : "";
		const label = `${frappe.utils.escape_html(row.label)}${who} &middot; ${frappe.datetime.str_to_user(
			row.from_time
		)}`;

		return `
			<tr class="pb-detail-row ${pending === undefined ? "" : "pb-dirty"}" data-name="${row.name}">
				<td class="pb-work" style="padding-left: ${12 + row.indent * 20}px">
					<a href="/app/timesheet/${encodeURIComponent(row.timesheet)}">${label}</a>
					${row.locked ? `<span class="pb-lock" title="${__("Invoiced")}">&#128274;</span>` : ""}
				</td>
				<td class="pb-num">${format_number(row.hours, null, 2)}</td>
				<td class="pb-num pb-editable">${billing_cell}</td>
				<td class="pb-num">${format_number(non_billable, null, 2)}</td>
				<td class="pb-read">
					<input type="checkbox" class="pb-read-check" data-name="${row.name}"
						${row.mark_as_read ? "checked" : ""}>
				</td>
			</tr>
		`;
	}

	has_children(task_row) {
		const index = this.rows.indexOf(task_row);
		const next = this.rows[index + 1];
		return !!next && next.indent > task_row.indent;
	}

	bind_row_events() {
		this.body.on("click", ".pb-task-row .pb-caret", (event) => {
			const name = $(event.currentTarget).closest("tr").data("name");
			if (this.collapsed.has(name)) {
				this.collapsed.delete(name);
			} else {
				this.collapsed.add(name);
			}
			this.render_rows();
		});

		this.body.on("change", ".pb-input", (event) => {
			const input = $(event.currentTarget);
			const name = input.data("name");
			const row = this.rows.find((r) => r.name === name);
			const value = flt(input.val());

			// Checked here as well as on the server so a typo is caught before it
			// costs a round trip; update_billing_hours re-checks both bounds because
			// this method is whitelisted and reachable without the page.
			if (value < 0 || value > flt(row.hours)) {
				frappe.show_alert({
					message: __("Billable hours must be between 0 and {0}.", [row.hours]),
					indicator: "red",
				});
				input.val(format_number(row.billing_hours, null, 2));
				return;
			}

			if (value === flt(row.billing_hours)) {
				delete this.pending[name];
			} else {
				this.pending[name] = value;
			}

			this.set_dirty(Object.keys(this.pending).length > 0);
			this.render_rows();
		});

		this.body.on("change", ".pb-read-check", (event) => {
			const input = $(event.currentTarget);
			const name = input.data("name");
			const value = input.is(":checked") ? 1 : 0;

			frappe
				.call({
					method: this.method("set_mark_as_read"),
					args: { names: [name], value: value },
				})
				.then(() => {
					const row = this.rows.find((r) => r.name === name);
					if (row) row.mark_as_read = value;
				});
		});
	}

	set_dirty(dirty) {
		this.page.btn_primary.toggleClass("hide", !dirty);
	}

	save() {
		const updates = Object.keys(this.pending).map((name) => ({
			name: name,
			billing_hours: this.pending[name],
		}));

		if (!updates.length) return;

		frappe
			.call({
				method: this.method("update_billing_hours"),
				args: { updates: updates },
				freeze: true,
				freeze_message: __("Updating billable hours..."),
			})
			.then((r) => {
				if (!r.message) return;
				frappe.show_alert({
					message: __("{0} row(s) updated.", [r.message.updated]),
					indicator: "green",
				});
				// Reloaded rather than patched in place: one edit moves the whole
				// chain of rollups above it, so every ancestor row on screen is stale.
				this.refresh();
			});
	}
}
